CIP-ledger-evolution/README.md

```diff
@@ -0,0 +1,302 @@
+
+## Abstract
+
+This CIP proposes a process for proposing changes to the Cardano ledger.
+
+## Motivation
+
+The ledger is responsible for processing transactions and updating the shared state of the network.
+It also processes block headers and handles the state transformation from one epoch to the next
+(e.g. computing the staking rewards).
+It is a very user-facing component;
+adding new functionality to the Cardano network often involves adding features to the ledger.
+
+Most of the state maintained by the ledger relates to the
+[Extended UTxO accounting model](https://iohk.io/en/research/library/papers/the-extended-utxo-model/) and
+support for [Ouroboros](https://iohk.io/en/research/library/papers/ouroboros-a-provably-secure-proof-of-stake-blockchain-protocol/),
+the proof-of-stake consensus mechanism used in Cardano.
+
+This CIP aims to give guidance for future CIPs related to the ledger,
+making it a registered category of the CIP process[^1].
+
+Many thanks Michael Peyton Jones to writing [CIP-35] and forging this path for us.
+Some of the language here has been taken directly from it.
+
+## Background
+
+### Terminology
+
+Context for the terminology used in this document is given in [CIP-59].
+
+### Specifications
+
+The ledger is specified as a state transition system using a
+[small step operational semantics](STS.md).
+We refer to this framework as the *Small Step Semantics for Cardano*, or the *STS* for short.
+An understanding of the existing STS specifications for the
+[existing ledger eras](https://github.com/input-output-hk/cardano-ledger#cardano-ledger)
+is often required to fully understand the implications of changes to the ledger
+(though an understanding of the Haskell implementation is a fair substitute).
+
+The STS framework leaves both cryptographic primitives and the serialization format abstract.
+These specifications need to be complete enough to realize a full implementation of the ledger
+given the cryptographic primitives and serialization format.
+The cryptographic primitives are described as appendices to the STS specifications,
+and the serialization format is given as a
+[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).
+There SHOULD be one STS specification per ledger era
+(the Allegra ledger era is currently combined with the Mary era).
+
+From the Byron to the Babbage ledger eras, the STS frameworks were written in $\LaTeX$.
+Starting in the Conway ledger era, literate Agda will be used.
+During the transition from $\LaTeX$ to literate Agda, we will take advantage
+of the ability to substitute $\LaTeX$ in place of Agda code when needed for expedience.
+With time, the Agda specification will gloriously blur the line between specification
+and reference implementation,
+just as the type system upon which Agda is built blurs the line between proof and program.
+
+### Ledger eras
+
+A ledger era is a collection of features added to the ledger which are introduced at a hard fork.
```

Comment:
I've struggled to make this precise. It's a bundle of features, and we bundle things mostly according to some vague coherence between features and release schedules. Intra-era hard forks are mostly a hack that we try to avoid.

Reply:
I think it's fine for the bundling to be ad-hoc, my question was rather what kind of changes have to go in _some_ new era. 

> Intra-era hard forks are mostly a hack that we try to avoid.

Why, OOI? For Plutus it seems relatively clear that there are some changes that it's safe to make on a protocol version bump without a new ledger era, i.e. in an intra-era HF (even if they may get released in a bigger release). I'm not sure why it's so much sketchier for the ledger? I guess there's just more tricky details!

Reply:
> what kind of changes have to go in some new era

It could be my lack of imagination, but I don't have a structured answer to give.

> tricky details

Let me see if I can give a summary that is coherent. :sweat_smile: . Many of our types and functions and rules are parametrized by an uninhabited `era` type. We also use a lot of type families, so that, for instance, `TxOut era` is abstract, and you need to know, say, that the Alonzo era uses the `AlonzoTxOut era` type. The name "Alonzo" here doesn't mean that it is used _only_ in the Alonzo era, but that it was _introduced_ in the Alonzo era. In this respect, our code very much mimics our "diff specs". Moreover, all the Alonzo code is in a package call `cardano-ledger-alonzo`. There are also type classes which provide functions across eras. So this is usually what is in our head when we are looking at code and thinking about how things work in a given era. In walks the `HardForks` and `SoftForks` module, which allows you to switch behavior based on being above or below a given major protocol version. So you might be in a function in the Shelley package, which is used by Alonzo, and you change the behavior in a very specific and localized way that only manifests in, say, the first Mary protocol version. When used like a surgical device, for a very small an isolated change, it saves an incredible amount of time and code copy (or added parametricity which isn't otherwise needed, etc). But I think the whole team would agree that if we used this too much, the code would become too hard to reason about.